modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexMaxConcurrentTasksTests.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.reindex;
+
+import org.elasticsearch.action.admin.cluster.settings.ClusterUpdateSettingsRequest;
+import org.elasticsearch.action.index.IndexRequestBuilder;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
+
+public class ReindexMaxConcurrentTasksTests extends ReindexTestCase {
+
+    public void testMaxConcurrentTasks() throws Exception {
+        List<IndexRequestBuilder> docs = new ArrayList<>();
+        int max = between(150, 500);
+        for (int i = 0; i < max; i++) {
+            docs.add(client().prepareIndex("source").setId(Integer.toString(i)).setSource("foo", "a"));
+        }
+
+        indexRandom(true, docs);
+        assertHitCount(client().prepareSearch("source").setSize(0).get(), max);
+
+        try {
+            ClusterUpdateSettingsRequest settingsUpdate = new ClusterUpdateSettingsRequest();
+            Settings settings = Settings.builder().put(TransportStartReindexTaskAction.MAX_CONCURRENT_REINDEX_TASKS.getKey(), 3).build();
+            settingsUpdate.persistentSettings(settings);
+            client().admin().cluster().updateSettings(settingsUpdate).get();
+
+            // Since we are only using validation on the data nodes and this test is inherently racy, we
```

Comment:
I think we should be able to just submit 3 good requests and then validate one (or several) failing requests afterwards? I think this is only racy in unhealthy setups, AFAIK, a cluster state update waits for all nodes to have applied the state before returning (with timeout).